app/src/main/cpp/WidgetBorder.cpp

```diff
@@ -0,0 +1,239 @@
+  if ((v_uv.x < 0.0f) || (v_uv.x > 1.0f)) {
+    color.a = 0.0f;
+  }
+  gl_FragColor = color * v_color;
+}
+
+)SHADER";
+
+struct WidgetBorder::State {
+  CylinderPtr cylinder;
+  vrb::GeometryPtr geometry;
+  vrb::TransformPtr transform;
+
+  template<typename T>
+  void UpdateMaterial(const T &aTarget, const vrb::Color &aDiffuse) {
+    vrb::Color ambient(0.5f, 0.5f, 0.5f, 1.0f);
+    aTarget->SetMaterial(ambient, aDiffuse, vrb::Color(0.0f, 0.0f, 0.0f), 0.0f);
+  }
+
+  void AppendBorder(const vrb::GeometryPtr &geometry, GLint index1, GLint index2,
+                    const vrb::Vector &offset, const vrb::Color &aColor, GLint &currentIndex) {
+    vrb::VertexArrayPtr array = geometry->GetVertexArray();
+    vrb::Vector offset1 = array->GetVertex(index1 - 1) + offset;
+    vrb::Vector offset2 = array->GetVertex(index2 - 1) + offset;
+    array->AppendVertex(offset1);
+    array->AppendVertex(offset2);
+    array->AppendColor(aColor);
+    array->AppendColor(aColor);
+
+    GLint index3 = currentIndex++;
+    GLint index4 = currentIndex++;
+    std::vector<int> index;
+    if (offset.y() > 0.0f) {
+      index = {index1, index2, index4, index3};
+    } else if (offset.y() < 0.0f) {
+      index = {index3, index4, index2, index1};
+    } else if (offset.x() > 0.0f) {
+      index = {index1, index3, index4, index2};
+    } else {
+      index = {index3, index1, index2, index4};
+    }
+
+    std::vector<int> normalIndex;
+    normalIndex.push_back(1);
+    normalIndex.push_back(1);
+    normalIndex.push_back(1);
+    normalIndex.push_back(1);
+    geometry->AddFace(index, index, normalIndex);
+  }
+
+  vrb::GeometryPtr
+  CreateGeometry(vrb::CreationContextPtr aContext, const vrb::Vector &aMin, const vrb::Vector &aMax,
+                 const device::EyeRect &aBorder) {
+    vrb::VertexArrayPtr array = vrb::VertexArray::Create(aContext);
+    vrb::Color solid(1.0f, 1.0f, 1.0f, 1.0f);
+    vrb::Color border(1.0f, 1.0f, 1.0f, 0.0f);
+
+    const vrb::Vector bottomRight(aMax.x(), aMin.y(), aMin.z());
+    array->AppendVertex(aMin); // Bottom left
+    array->AppendVertex(bottomRight); // Bottom right
+    array->AppendVertex(aMax); // Top right
+    array->AppendVertex(vrb::Vector(aMin.x(), aMax.y(), aMax.z())); // Top left
+    for (int i = 0; i < 4; ++i) {
+      array->AppendColor(solid);
+    }
+
+    vrb::Vector normal = (bottomRight - aMin).Cross(aMax - aMin).Normalize();
+    array->AppendNormal(normal);
+
+    vrb::RenderStatePtr state = vrb::RenderState::Create(aContext);
+    state->SetVertexColorEnabled(true);
+    vrb::GeometryPtr geometry = vrb::Geometry::Create(aContext);
+    geometry->SetVertexArray(array);
+    geometry->SetRenderState(state);
+
+    std::vector<int> index;
+    index.push_back(1);
+    index.push_back(2);
+    index.push_back(3);
+    index.push_back(4);
+    std::vector<int> normalIndex;
+    normalIndex.push_back(1);
+    normalIndex.push_back(1);
+    normalIndex.push_back(1);
+    normalIndex.push_back(1);
```

Comment:
We seem to do this, a lot, I wonder if we could refactor simple geometry creation out into a factory of some point?

Reply:
Actually, I see you sort of did that already here.